Add a secimtools suite

moskalenko opened the pull request:
This is a set of tools for metabolomics for a submitted publication.

bgruening commented:
This is an incredible contribution! Thanks a bunch @moskalenko!

@RJMW can you maybe also have a look at it :)

bgruening commented on tools/secimtools/.shed.yml:
```diff
@@ -0,0 +1,34 @@
+description: "Metabolomics tools from the SECIM project"
+homepage_url: https://github.com/secimTools
+long_description: |
+  "The SECIMTools is suite includes a comprehensive set of quality control
```
indentaion

moskalenko replied:
Done by @shiltemann

bgruening commented on tools/secimtools/.shed.yml:
```diff
@@ -0,0 +1,34 @@
+  name: "suite_secimtools"
+  description: "Metabolomics tools from the SECIM project"
+  long_description: |
+    "The SECIMTools is suite includes a comprehensive set of quality control
```
indentation

moskalenko replied:
Done by @shiltemann

bgruening commented on tools/secimtools/anova_fixed.xml:
```diff
@@ -0,0 +1,110 @@
+<tool id="anova_fixed" name="Analysis of Variance (ANOVA) Fixed Effects Model" version="1.0.0">
```
can we prefix all tool IDs with `secimtools_`

bgruening replied:
You might also want to macrofy the version and set it to the @WRAPPER_VERSION@?

moskalenko replied:
The tools are going to diverge as updates are added, probably better to keep the versions in wrappers, but that can be done once they begin diverging. I'll use the @WRAPPER_VERSION@ for the pull request. Thank you.

nsoranzo replied:
In that case you can use e.g.`@WRAPPER_VERSION@.1` .

bgruening commented on tools/secimtools/anova_fixed.xml:
```diff
@@ -0,0 +1,110 @@
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <requirements>
```
You probably want to use your requirement macro here.

moskalenko replied:
Moved up by @shiltemann

bgruening commented on tools/secimtools/anova_fixed.xml:
```diff
@@ -0,0 +1,110 @@
+        <requirement type="python-module">statsmodels</requirement>
+        <requirement type="python-module">matplotlib</requirement>
+    </requirements>
+    <command interpreter="python">
```
the interpreter tag is depreacted, simply remove it :)

bgruening commented on tools/secimtools/anova_fixed.xml:
```diff
@@ -0,0 +1,110 @@
+    </requirements>
+    <command interpreter="python">
+            anova_fixed.py
+            --input $input
```
all type=data, outputs and text parameters should be single quoted.

bgruening commented on tools/secimtools/anova_fixed.xml:
```diff
@@ -0,0 +1,110 @@
+    <inputs>
+        <param name="input" type="data" format="tabular" label="Wide Dataset" help="Input dataset in wide format and tab separated. If file is not tab separated see TIP below."/>
+        <param name="design" type="data" format="tabular" label="Design File" help="Design file tab separated. Note you need a 'sampleID' column. If not tab separated see TIP below."/>
+        <param name="uniqID" type="text" size="30" value="" label="Unique Feature ID" help="Name of the column in your Wide Dataset that has unique Feature IDs."/>
```
the size attribute can be removed it is not used anymore

bgruening commented on tools/secimtools/anova_fixed.xml:
```diff
@@ -0,0 +1,110 @@
+            --design $design
+            --ID $uniqID
+            --factors "$factor"
+            --factorTypes $factorTypes
```
single qutes

bgruening commented on tools/secimtools/bland_altman_plot.xml:
```diff
@@ -0,0 +1,134 @@
+    </tests>
+<help>
+
+@TIP_AND_WARNING@
```
This is not defined in macros.xml isn't it?

bgruening commented on tools/secimtools/blank_feature_filtering_flags.xml:
```diff
@@ -0,0 +1,115 @@
+        <import>macros.xml</import>
+    </macros>
+    <requirements>
+        <requirement type="python-module">numpy</requirement>
```
use macro

moskalenko replied:
Done by @shiltemann

shiltemann replied:
I did not do this yet, the macro file is imported, but to actually use it, this should be changed to something like:

```
<expand macro="requirements">
    <requirement type="package">numpy</requirement>
     ...
</expand>

bgruening commented on tools/secimtools/blank_feature_filtering_flags.xml:
```diff
@@ -0,0 +1,115 @@
+<tool id="blank_feature_filtering_flags" name="Blank Feature Filtering (BFF) Flag" version="2.1,0">
```
version looks wired and should use the macro

bgruening commented on tools/secimtools/compare_flags.xml:
```diff
@@ -0,0 +1,92 @@
+        <requirement type="python-module">pandas</requirement>
+    </requirements>
+    <command interpreter="python">compare_flags.py
+        --input $input
```
quotes

shiltemann commented on tools/secimtools/anova_fixed.xml:
```diff
@@ -0,0 +1,110 @@
+        <data format="pdf" name="qq_plots" label="${tool.name} on ${on_string}: QQ Plots"/>
+        <data format="pdf" name="volcano_plots" label="${tool.name} on ${on_string}: Volcano Plots"/>
+    </outputs>
+    <macros>
```
you have this macros section twice here (and in many other tools) please remove the second one and keep only the one  near the top

shiltemann commented on tools/secimtools/retention_time_flags.xml:
```diff
@@ -0,0 +1,118 @@
+    <requirement type="python-module">palettable</requirement>
+  </requirements>
+  <stdio>
+    <exit_code range="1:" level="warning" description="RuntimeWarning"/>
```
you can remove this `stdio` section and use `detect_errors="exit_code"` on the `command` tag (and consider doing this for other tools too maybe?)

shiltemann commented on tools/secimtools/bland_altman_plot.xml:
```diff
@@ -0,0 +1,134 @@
+    <requirement type="python-module">matplotlib</requirement>
+    <requirement type="python-module">palettable</requirement>
+  </requirements>
+  <command interpreter="python">
```
could you put all command and help sections in CDATA tags? (this allows presence of special characters in these sections), so like this: 

```
<command><![CDATA[
 
your command here

]]></command>
```

shiltemann commented on tools/secimtools/summarize_flags.xml:
```diff
@@ -0,0 +1,75 @@
+     <test>
+        <param name="flags"   value="ST000006_lasso_enet_var_select_flags.tsv"/>
+        <param name="uniqID"  value="Retention_Index" />
+        <output name="output" file="ST000006_summarize_flags_outSummary.tsv" />
```
there is no output named `output`, did you mean `summaryFlags`?

shiltemann commented:
(made a [PR](https://github.com/secimTools/tools-iuc/pull/1) to your branch with some of the more tedious changes I requested (indentation and such) )

moskalenko commented:
Thank you for the changes. I'm merging the changes and doing other suggested fixes and updates.

moskalenko commented:
Not sure why the Travis run is failing.

bgruening commented on tools/secimtools/macros.xml:
```diff
@@ -0,0 +1,14 @@
+        </requirements>
+    </xml>
+    <xml name="citations">
+        <citations>
```
@moskalenko do you want to add a citation here, maybe a preprint?

bgruening commented on tools/secimtools/magnitude_difference_flags.xml:
```diff
@@ -0,0 +1,105 @@
+        <import>macros.xml</import>
+    </macros>
+    <requirements>
+        <requirement type="python-module">numpy</requirement>
```
You should use the requirements from the macro here and btw. python-module is not used anymore, this should be package.

moskalenko commented:
@shiltemann Saskia, thank you so much for all the formatting fixes!

* I changed all wrappers to expand the requirements macro instead of defining requirements in each wrapper
* I added a citation (in press) as recommended.

moskalenko commented:
It looks like the failed tests were marked as such because of the matplotlib font warning:

"UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
|  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment."

shiltemann commented:
yes, it seems this warning message is written to stderr, which by default will make Galaxy think an error occurred (even if it didn't) ..if you use `<command detect_errors="exit_code">` the exit code of the command will be used to determine success/failure rather than non-emptiness of stderr, and should help with these false errors

bgruening commented:
We have also this problem `|  /home/travis/conda/envs/__secimtools@1.0.0/bin/bland_altman_plot.py: line 30: syntax error near unexpected token `'Agg''`

moskalenko commented:
In bland_altman_plot.py we set the Agg backend after importing matlab because of matlab errors in headless runs. 

  28 import matplotlib
  29 import pandas as pd
  30 matplotlib.use('Agg')

bgruening commented:
@moskalenko that's usually fine, the error needs to be elsewhere. Can you add the detect_error things and we see if most of the tools are passing. No idea yet about the Agg stuff.

shiltemann commented:
I think the Agg error may be due to a stray space at beginning of shebang line in that script? https://github.com/secimTools/SECIMTools/blob/master/src/scripts/bland_altman_plot.py#L1

shiltemann commented on tools/secimtools/bland_altman_plot.xml:
```diff
@@ -0,0 +1,124 @@
+        <param name="design" type="data" format="tabular" label="Design File" help="Design file tab separated. Note you need a 'sampleID' column. If not tab separated see TIP below."/>
+        <param name="uniqID" size="30" type="text" value="" label="Unique Feature ID" help="Name of the column in your Wide Dataset that has unique feature IDs."/>
+        <param name="resid_cutoff" type="text" size="30" value="3" label="Outlier Cutoff" help="Residual cutoff value, this value will flag samples with residuals ≥ this cutoff value."/>
+        <param name="sample_cutoff" type="text" size="30" value="0.2" label="Sample Flag Cutoff" help="Flag a sample as 1 if the proportion of features within a sample that are outliers exceeds this cutoff. [Number between 0-1]."/>
```
help text says "number between 0-1", so this should be changed to `type="float" min="0" max="1"` (and there are more tools with text params that can be changed to float/integer type)

moskalenko commented:
It looks like the remaining errors are due to the tests not able to deal with the matplotlib setting Agg backend for headless execution. I'm not sure how to address this in the context of TravisCI. These runs will fail in a real run on a compute cluster with headless nodes if the backend setter is removed.

shiltemann commented:
@moskalenko did you see my PR to your secimtools repo? there is a space at the beginning of your shebang line in that script causing the python interpreter to not be found (there were a bunch more strange errors before the Agg error due to this) ..when I fixed this line locally it worked again, so if you could merge that and update the conda package I think it should fix the problem

shiltemann commented:
also, it appears that the test for `mzrt_match` is hitting the 20-minute timeout in travis, is there any way you could tweak the parameters and/or input data to speed this test up you think?

moskalenko commented:
I'll get a smaller dataset for the mzrt_match test. I'll be happy to update the bland_altmann script to remove a space in shebang. I've already uploaded a new build to pypi. However, I also see a lot of tests failing/segfaulting with something like 'cannot find X11'.

shiltemann commented:
oh, I'm not seeing the X11 errors, for which tools does that happen? I know it is definitely possible in travis with the Agg backend as I use the same in my mothur tools (but there I had to do some env variable/conf file magic so I could see if same works here)

moskalenko commented:
I've seen those errors in Travis logs.

shiltemann commented:
could you give a link to a specific line in travis? searching for X11 didn't turn anything up for me (only something similar in bland_altman script, but think that may all be due to missing interpreter, is it happening in other tools too?)

shiltemann commented:
ah nvm, see it happens in `distribution_features` too, but in that script the shebang line is missing altogether so could be same cause?

moskalenko commented:
Segfaults, which don't happen locally:

--------------------------------------
History with id 683bc220e21425bb in error - summary of datasets in error below.
--------------------------------------
| 3 - Scatter Plot 3D on data 2 and data 1: scatter plots (HID - NAME) 
| Dataset Blurb:
|  error
| Dataset Info:
|  2017-07-28 22:24:45,173 - INFO - Using Tableau_20 color scheme from tableau palette
|  Fatal error: Exit code 134 ()
|  /home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building t
| Dataset Job Standard Output:
|  2017-07-28 22:24:45,173 - INFO - Using Tableau_20 color scheme from tableau palette
| Dataset Job Standard Error:
|  Fatal error: Exit code 134 ()
|  /home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
|  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
|  QXcbConnection: Could not connect to display
|  /tmp/tmpAtfyvc/job_working_directory/000/20/tool_script.sh: line 25:  6923 Aborted                 (core dumped) scatter_plot_3D.py --input /tmp/tmpAtfyvc/files/000/dataset_23.dat --ID sampleID --X PC1 --Y PC2 --Z PC3 --figure /tmp/tmpAtfyvc/files/000/dataset_25.dat --design /tmp/tmpAtfyvc/files/000/dataset_24.dat --group White_wine_type_and_source

shiltemann commented:
ah thanks :)

moskalenko commented:
Sorry, I missed your PR against SECIMTools, but I remembered your comment, so I fixed that shebang. I should do another build to fix/add shebangs to all scripts and do a pull request in conda to update the package. However, I'd like to see at least some successful tests with matplotlib-using scripts first to make sure.

shiltemann commented:
ok, this "unable to find display" error seems to happen for tools `random_forest`, `mahanalobis_distance`, `scatter_plot_3D`, `scatter_plot_2D` ..I don't see lines setting the Agg backend in these scripts, should that be added or..?

shiltemann commented:
(fwiw `hierarchical_clustering_heatmap` test passes and uses matplotlib it looks like?)

bgruening commented on tools/secimtools/.travis.yml:
```diff
@@ -0,0 +1,14 @@
+language: python
+
+python:
```
@moskalenko this is not a good idea. We need to get your tools to work without this, because you can not assume an admin is doing the same on the cluster.

Python plotting should work, deeptools and many other tools are doing it as well without these modifications.
With which tool do you need help?

moskalenko replied:
I was just exploring approaches other projects used to get around the matplotlib issue. If that addition is not a good idea please patch it out, no problem.

I'm just about ready to push an updated python package to pypi and will submit a bioconda pull request for the package update afterwards, just waiting for tests to finish. The updated package has shebang, matplotlib backend, and assorted code formatting fixes.

bgruening replied:
Sounds great!

moskalenko commented:
All kinds of issues - matplotlib DISPLAY issues even though we have addressed that in .travis.yml

These are new I think. It looks like conda dependencies or the Travis env have problems:

 File "/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/pymc/utils.py", line 14, in <module>
|  from . import flib
|  ImportError: libopenblas.so.0: cannot open shared object file: No such file or directory

I am not sure why, for example, mahalanobis_distance failed as it's set to 'detect_errors="exit_code"' and the only issue I see there is a deprecation warning in stderr.

begin tool stderr << -----------------------
/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/secimtools/visualManager/manager_figure.py:307: MatplotlibDeprecationWarning: The set_axis_bgcolor function was deprecated in version 2.0. Use set_facecolor instead.
  self.ax[0].set_axis_bgcolor('w')

moskalenko commented:
I wonder what's gong on with icu-56.1

| Dataset Job Standard Error:
|  Fatal error: Exit code 1 ()
|  Traceback (most recent call last):
|  File "/home/travis/conda/envs/__secimtools@1.0.0/bin/magnitude_difference_flags.py", line 26, in <module>
|  from lxml import etree
|  ImportError: libicui18n.so.58: cannot open shared object file: No such file or directory

moskalenko commented:
This is the prevalent error now:

| Dataset Job Standard Error:
|  Fatal error: Exit code 1 ()
|  Traceback (most recent call last):
|  File "/home/travis/conda/envs/__secimtools@1.0.0/bin/random_forest.py", line 262, in <module>
|  main(args)
|  File "/home/travis/conda/envs/__secimtools@1.0.0/bin/random_forest.py", line 230, in main
|  plotVarImportance(data=df_importance, pdf=pdfOut, var=args.num)
|  File "/home/travis/conda/envs/__secimtools@1.0.0/bin/random_forest.py", line 122, in plotVarImportance
|  fh = figureHandler(proj='2d', figsize=(8,8))
|  File "/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/secimtools/visualManager/manager_figure.py", line 298, in __init__
|  self.fig = plt.figure(figsize=figsize)
|  File "/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/matplotlib/pyplot.py", line 535, in figure
|  **kwargs)
|  File "/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt4agg.py", line 46, in new_figure_manager
|  return new_figure_manager_given_figure(num, thisFig)
|  File "/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt4agg.py", line 53, in new_figure_manager_given_figure
|  canvas = FigureCanvasQTAgg(figure)
|  File "/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt4agg.py", line 76, in __init__
|  FigureCanvasQT.__init__(self, figure)
|  File "/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt4.py", line 63, in __init__
|  _create_qApp()
|  File "/home/travis/conda/envs/__secimtools@1.0.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt5.py", line 136, in _create_qApp
|  raise RuntimeError('Invalid DISPLAY variable')
|  RuntimeError: Invalid DISPLAY variable

The Agg backend should be used according to the travis config.

bgruening commented:
@moskalenko it still seems that the QT backend is used. What do you mean by this was addressed in travis?

moskalenko commented:
Rebasing and running planemo tests to make a push to resolve the last conflicts while at Galaxy ConFest 2018.

moskalenko commented:
Rebased

nsoranzo commented:
@moskalenko I think because this [change](https://github.com/bioconda/bioconda-recipes/commit/bc2814c71631b71d683fda639be161bd2bcf9b0d#diff-e72e6dc54e18916ebea0282eb3a91bfcL16) made the bioconda package available for all Python versions. Ping @bgruening

bgruening commented:
This still seems to be broken for me. For example I get this error locally.

```python
  from scipy.stats.stats import ss
  ImportError: cannot import name ss
```

And I found this: `scipy.stats.stats.ss has been deprecated since scipy v0.17`
